File: LeetCode/LeetCode 1492. n 的第 k 个因子.cpp
class Solution {
public:
    int kthFactor(int n, int k) {
        vector <int> ss;
        ss.push_back(-1);
        int now = 0, flag = 0;
        for(int i = 1; i * i <= n; i++) {
            if(n % i == 0) {
                now++;
                ss.push_back(i);
                if(now == k) return i;
            }
            if(i * i == n) flag = 1;
        }
        k -= now;
        int index = now - k + 1 - flag;
        return index == 0 ? -1 : n / ss[index];
    }
};

File: LeetCode/LeetCode 1492. n 的第 k 个因子.py
class Solution:
    def kthFactor(self, n: int, k: int) -> int:
        ss = [-1]
        now = flag = 0
        for i in range(1, n + 1):
            if i * i > n:
                break
            elif i * i == n:
                flag = 1

            if n % i == 0:
                now += 1
                ss.append(i)
                if now == k:
                     return i
        
        k -= now
        index = now - k + 1 - flag
        if index == 0:
            return -1
        else:
            return n // ss[index]